tasks/Task.cpp

```diff
@@ -28,27 +28,265 @@ bool Task::configureHook()
+{
+    //  Compute the initial pose
+    pose.clearPoint();
+
+    for (int i=0 ; i < 4 ; i++)
+    {
+        base::Vector2d aux;
+        aux = corners[i];
+        //vpImagePoint is instanciated using (v,u) instead of (u,v)
+        vpImagePoint a(aux[1],aux[0]);
+        double x,y;
+        vpPixelMeterConversion::convertPoint(cam_cal,a,x,y);
+        point[i].set_x(x);
+        point[i].set_y(y);
+        pose.addPoint(point[i]) ;
+    }
+
+    pose.computePose(vpPose::LAGRANGE, cMo);
+    pose.computePose(vpPose::VIRTUAL_VS, cMo);
+
+    // Update Feature 1 - p
+    // Sets the current position of the visual feature
+    P.track(cMo);
+    vpFeatureBuilder::create(p, P);
+
+    //////////////////////////////
+    // Update Feature 2 - depth //
+    //////////////////////////////
+
+    //Takes the features, calculate the intersection of the diagonals of the square and
+    //given the camera parameters, retrieve the coordinates in meters.
+    vpImagePoint refP;
+    //double x,y;
+    vpImagePoint cog[4];
+
+    for (int i=0; i< 4; i++)
+    {
+        base::Vector2d aux;
+        aux = corners[i];
+        //according to vpImagePoint documentation (i,j) is equivalent to (v,u)
+        cog[i].set_ij(aux[1], aux[0]);
+    }
+
+    // Compute the coordinates of the 2D reference point from the
+    // intersection of the diagonals of the square
+    double a1 = (cog[0].get_i()- cog[2].get_i()) /
+        (cog[0].get_j()- cog[2].get_j());
+    double a2 = (cog[1].get_i()- cog[3].get_i()) /
+        (cog[1].get_j()- cog[3].get_j());
+    double b1 = cog[0].get_i() - a1*cog[0].get_j();
+    double b2 = cog[1].get_i() - a2*cog[1].get_j();
+
+    refP.set_j( (b1-b2) / (a2-a1) );
+    refP.set_i( a1*refP.get_j() + b1 );
+    double x,y;
+    vpPixelMeterConversion::convertPoint(cam_cal, refP, x, y);
+
+    // Update the 2D ref point visual feature
+    double Z;
+    Z = P.get_Z();
+    p.set_xyZ(x, y, Z);
+    depth.buildFrom(P.get_x(), P.get_y(), Z, log(Z/Zd));
+
+    //////////////////////////////
+    // Update Feature 3 - t, tu //
+    //////////////////////////////
+    cdMc = cdMo*cMo.inverse();
+    t.buildFrom(cdMc);
+    tu.buildFrom(cdMc) ;
+
+}
+
+bool Task::updateDesiredPose()
+{
+    // Initialise a desired pose to compute s*, the desired 2D point features
+    base::LinearAngular6DCommand setpoint;
+    if (_cmd_in.read(setpoint) == RTT::NewData)
+    {
+        vpTranslationVector cdto(setpoint.linear[0], setpoint.linear[1], setpoint.linear[2]);
+        vpRxyzVector cdro(setpoint.angular[0], setpoint.angular[1], setpoint.angular[2]);
+
+        //Build the rotation and homogeneus matrix according to desired pose
+        vpRotationMatrix cdRo(cdro); // Build the rotation matrix
+        cdMo.buildFrom(cdto, cdRo); // Build the homogeneous matrix
+
+        P.track(cdMo);
+        vpFeatureBuilder::create(pd, P);
+        Zd = P.get_Z();
+    }
+    else
+    {
+        exception(MISSING_SETPOINT);
```

Comment:
This is horribly wrong:
 - at initialization time, this is a normal situation. All inputs are asynchronous, and unordered.
 - after initialization, it is *also* normal because you compare against NewData

Read the port within updateHook and pass the pose to the updateDesiredPose(...) method. Generally speaking, refactor to keep all the logic about port reading in updateHook, to avoid "hiding" it deeper.

Also, never do

```
myMethod() {
   if (ok_condition)
   {
       very long piece of code
       return true;
   }
   else (failure_condition)
   {
       return false;
   }
}
```

The following is a lot more readable:

myMethod() {
   if (failure_condition)
      return false;

   very long piece of code
   return true;
}